File: game.py
import streamlit as st
import asyncio
import random

# Constants
WIDTH, HEIGHT = 800, 600
PLAYER_SIZE = 20
BALL_SIZE = 15
BULLET_SIZE = 5
PLAYER_SPEED = 5
BULLET_SPEED = 10
GOAL_SCORE = 5

# Initialize game state
if "players" not in st.session_state:
    st.session_state.players = [
        {"x": 100, "y": HEIGHT // 2, "color": "blue", "score": 0, "bullets": []},
        {"x": WIDTH - 100, "y": HEIGHT // 2, "color": "red", "score": 0, "bullets": []},
    ]
if "ball" not in st.session_state:
    st.session_state.ball = {"x": WIDTH // 2, "y": HEIGHT // 2, "color": "yellow"}
if "game_over" not in st.session_state:
    st.session_state.game_over = False

# Helper functions
def move_player(player, dx, dy):
    player["x"] = max(PLAYER_SIZE, min(player["x"] + dx, WIDTH - PLAYER_SIZE))
    player["y"] = max(PLAYER_SIZE, min(player["y"] + dy, HEIGHT - PLAYER_SIZE))

def shoot_bullet(player):
    target_x = random.randint(0, WIDTH)
    target_y = random.randint(0, HEIGHT)
    bullet = {
        "x": player["x"],
        "y": player["y"],
        "target_x": target_x,
        "target_y": target_y,
        "color": player["color"],
    }
    player["bullets"].append(bullet)

def move_bullets():
    for player in st.session_state.players:
        for bullet in player["bullets"]:
            dx = bullet["target_x"] - bullet["x"]
            dy = bullet["target_y"] - bullet["y"]
            dist = (dx**2 + dy**2) ** 0.5
            if dist == 0:
                dist = 1
            bullet["x"] += (dx / dist) * BULLET_SPEED
            bullet["y"] += (dy / dist) * BULLET_SPEED

def check_collisions():
    ball = st.session_state.ball
    for player in st.session_state.players:
        # Check if ball collides with player
        dist = ((ball["x"] - player["x"]) ** 2 + (ball["y"] - player["y"]) ** 2) ** 0.5
        if dist < BALL_SIZE + PLAYER_SIZE:
            ball["x"] = WIDTH // 2
            ball["y"] = HEIGHT // 2
            player["score"] += 1
            if player["score"] >= GOAL_SCORE:
                st.session_state.game_over = True

# Streamlit UI
st.title("Soccer Shooter Game")
st.write("Use the buttons to move and shoot!")

# Player controls
col1, col2 = st.columns(2)
with col1:
    if st.button("Move Player 1 Left"):
        move_player(st.session_state.players[0], -PLAYER_SPEED, 0)
    if st.button("Move Player 1 Right"):
        move_player(st.session_state.players[0], PLAYER_SPEED, 0)
    if st.button("Move Player 1 Up"):
        move_player(st.session_state.players[0], 0, -PLAYER_SPEED)
    if st.button("Move Player 1 Down"):
        move_player(st.session_state.players[0], 0, PLAYER_SPEED)
    if st.button("Shoot Player 1"):
        shoot_bullet(st.session_state.players[0])

with col2:
    if st.button("Move Player 2 Left"):
        move_player(st.session_state.players[1], -PLAYER_SPEED, 0)
    if st.button("Move Player 2 Right"):
        move_player(st.session_state.players[1], PLAYER_SPEED, 0)
    if st.button("Move Player 2 Up"):
        move_player(st.session_state.players[1], 0, -PLAYER_SPEED)
    if st.button("Move Player 2 Down"):
        move_player(st.session_state.players[1], 0, PLAYER_SPEED)
    if st.button("Shoot Player 2"):
        shoot_bullet(st.session_state.players[1])

# Game canvas
canvas = st.empty()

# Game loop
async def game_loop():
    while not st.session_state.game_over:
        move_bullets()
        check_collisions()

        # Draw game state
        canvas.markdown(
            f"""
            <div style="position: relative; width: {WIDTH}px; height: {HEIGHT}px; background: black;">
                <div style="position: absolute; left: {st.session_state.ball['x']}px; top: {st.session_state.ball['y']}px; width: {BALL_SIZE}px; height: {BALL_SIZE}px; background: {st.session_state.ball['color']}; border-radius: 50%;"></div>
                {''.join(
                    f'<div style="position: absolute; left: {player["x"]}px; top: {player["y"]}px; width: {PLAYER_SIZE}px; height: {PLAYER_SIZE}px; background: {player["color"]}; border-radius: 50%;"></div>'
                    for player in st.session_state.players
                )}
                {''.join(
                    f'<div style="position: absolute; left: {bullet["x"]}px; top: {bullet["y"]}px; width: {BULLET_SIZE}px; height: {BULLET_SIZE}px; background: {bullet["color"]}; border-radius: 50%;"></div>'
                    for player in st.session_state.players for bullet in player["bullets"]
                )}
            </div>
            <p>Player 1 Score: {st.session_state.players[0]["score"]}</p>
            <p>Player 2 Score: {st.session_state.players[1]["score"]}</p>
            """,
            unsafe_allow_html=True,
        )

        await asyncio.sleep(0.1)

    st.write(f"Game Over! Player {1 if st.session_state.players[0]['score'] >= GOAL_SCORE else 2} wins!")

# Run the game loop
asyncio.run(game_loop())

File: game.js
const canvas = document.getElementById("gameCanvas");
const ctx = canvas.getContext("2d");

// Constants
const WIDTH = 800;
const HEIGHT = 600;
const WHITE = "#FFFFFF";
const RED = "#FF0000";
const GREEN = "#00FF00";
const BLUE = "#0000FF";
const YELLOW = "#FFFF00";
const BLACK = "#000000";
const FRICTION = 0.90;
const BALL_FRICTION = 0.97;
const BULLET_SPEED = 8;
const PLAYER_SPEED = 3;
const BULLET_DAMAGE = 20;
const FREEZE_TIME = 180;
const GOAL_SCORE = 5;
const BALL_IMPACT_MULTIPLIER = 1.2;
const BALL_BOUNCE = 0.8;
const DAMAGE_DECAY = 0.01;
const AUTO_AIM_PLAYER = 1;
const AUTO_AIM_BALL = 2;
const NO_AUTO_AIM = 0;

// Player Class
class Player {
  constructor(x, y, color, keys, autoAim = NO_AUTO_AIM) {
    this.x = x;
    this.y = y;
    this.radius = 20;
    this.color = color;
    this.velocity = { x: 0, y: 0 };
    this.health = 100;
    this.maxHealth = 100;
    this.bullets = [];
    this.score = 0;
    this.frozen = 0;
    this.keys = keys;
    this.autoAim = autoAim;
  }

  move() {
    if (this.frozen > 0) {
      this.frozen--;
      if (this.frozen === 0) this.health = this.maxHealth;
      return;
    }

    if (keys[this.keys.left]) this.velocity.x -= PLAYER_SPEED;
    if (keys[this.keys.right]) this.velocity.x += PLAYER_SPEED;
    if (keys[this.keys.up]) this.velocity.y -= PLAYER_SPEED;
    if (keys[this.keys.down]) this.velocity.y += PLAYER_SPEED;

    this.velocity.x *= FRICTION;
    this.velocity.y *= FRICTION;
    this.x += this.velocity.x;
    this.y += this.velocity.y;

    // Boundary checks
    this.x = Math.max(this.radius, Math.min(this.x, WIDTH - this.radius));
    this.y = Math.max(this.radius, Math.min(this.y, HEIGHT - this.radius));
  }

  shoot(target, ballPos = null, players = []) {
    if (this.frozen > 0) return;

    if (this.autoAim === AUTO_AIM_PLAYER && players.length > 0) {
      let closestPlayer = null;
      let closestDist = Infinity;
      for (const otherPlayer of players) {
        if (otherPlayer !== this) {
          const dx = this.x - otherPlayer.x;
          const dy = this.y - otherPlayer.y;
          const dist = Math.sqrt(dx * dx + dy * dy);
          if (dist < closestDist) {
            closestDist = dist;
            closestPlayer = otherPlayer;
          }
        }
      }
      if (closestPlayer) target = { x: closestPlayer.x, y: closestPlayer.y };
    } else if (this.autoAim === AUTO_AIM_BALL && ballPos) {
      target = ballPos;
    }

    const angle = Math.atan2(target.y - this.y, target.x - this.x);
    this.bullets.push(new Bullet(this.x, this.y, this.color, angle));
  }

  draw() {
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
    ctx.fillStyle = this.color;
    ctx.fill();
    ctx.closePath();

    // Health bar
    ctx.fillStyle = RED;
    ctx.fillRect(this.x - 20, this.y - 30, 40, 5);
    ctx.fillStyle = GREEN;
    ctx.fillRect(this.x - 20, this.y - 30, (40 * this.health) / this.maxHealth, 5);

    // Score
    ctx.fillStyle = WHITE;
    ctx.font = "16px Arial";
    ctx.fillText(`Score: ${this.score}`, this.x - 20, this.y - 45);
  }
}

// Bullet Class
class Bullet {
  constructor(x, y, color, angle) {
    this.x = x;
    this.y = y;
    this.radius = 4;
    this.color = color;
    this.velocity = {
      x: Math.cos(angle) * BULLET_SPEED,
      y: Math.sin(angle) * BULLET_SPEED,
    };
  }

  move() {
    this.x += this.velocity.x;
    this.y += this.velocity.y;
  }

  draw() {
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
    ctx.fillStyle = this.color;
    ctx.fill();
    ctx.closePath();
  }
}

// Ball Class
class Ball {
  constructor(x, y, color) {
    this.x = x;
    this.y = y;
    this.radius = 15;
    this.color = color;
    this.velocity = { x: 0, y: 0 };
  }

  move() {
    this.x += this.velocity.x;
    this.y += this.velocity.y;
    this.velocity.x *= BALL_FRICTION;
    this.velocity.y *= BALL_FRICTION;

    // Boundary checks
    if (this.x <= this.radius || this.x >= WIDTH - this.radius) this.velocity.x *= -BALL_BOUNCE;
    if (this.y <= this.radius || this.y >= HEIGHT - this.radius) this.velocity.y *= -BALL_BOUNCE;
  }

  draw() {
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
    ctx.fillStyle = this.color;
    ctx.fill();
    ctx.closePath();
  }
}

// Initialize Game Objects
const players = [
  new Player(100, HEIGHT / 2, BLUE, { left: "KeyA", right: "KeyD", up: "KeyW", down: "KeyS", shoot: "Space" }, AUTO_AIM_PLAYER),
  new Player(WIDTH - 100, HEIGHT / 2, RED, { left: "ArrowLeft", right: "ArrowRight", up: "ArrowUp", down: "ArrowDown", shoot: "Enter" }, AUTO_AIM_BALL),
];
const ball = new Ball(WIDTH / 2, HEIGHT / 2, YELLOW);
const goalposts = [
  { x: 0, y: HEIGHT / 3, width: 10, height: HEIGHT / 3 },
  { x: WIDTH - 10, y: HEIGHT / 3, width: 10, height: HEIGHT / 3 },
];

// Key State
const keys = {};

// Touch Controls
document
